### ISSUE TITLE
Use XGBoost classifier

### ISSUE THREAD

**Issue by DTTerastar** [ESPresense/ad-espresense-ips#5]
https://www.hackster.io/home_skillet/alexa-where-are-my-keys-aa0e82

**Comment by mjcumming** [comment]
This looks like a great way to solve for location in 3D. I have a few more ESP32's on the way and will try and see if there is a way to harvest data.

**Comment by iamandymcinnes** [comment]
With the issues you mentioned about python versions, what about moving the logic to the esp32s? Something like this: https://github.com/eloquentarduino/micromlgen

**Comment by DTTerastar** [comment]
Cat localizer is very cool.  The problem w/ one esp doing the ML is we need all the data from all nodes.

**Comment by Appyx** [comment]
My proposal: 
Why not use a docker container already prepared for scikit or XGBoost. This could run anywhere since only a mqtt python lib is necessary to communicate. When everything is working, we could just restart the container with a parameter so that it skips the training phase.
This can then easily be transformed into an addon which runs inside HomeassistantOS. (Not sure if training on ARM is easy tbh)

In production the user could simply switch the addon to training-mode and send the labels (rooms) via mqtt to the container.
When the user is finished he switches the addon back to non-training mode.

I guess I will try this out with scikit and a simple random forest.
Should be straight forward:
rssi1 rssi2 rssi3 rssi4 rssi5... = room1
rssi1 rssi2 rssi3 rssi4 rssi5... = room2
...

My only issue right now is that I might have to wait 10 seconds for all readings to arrive since the ESPs do not send readings at the same time. Can we improve the 5 second delay per reading?

**Comment by DTTerastar** [comment]
Very cool stuff.  I think that is a good approach.  Using appdaemon leads to pain.  I was thinking adding a select dropdown of what room the device is in, then when you switch it to another room use that as a new training point.

**Comment by Appyx** [comment]
I like the idea of a dropdown. Maybe even the already defined rooms in Home Assistant could be utilized somehow in a polished version.

However, I realized after posting my comment that this does not solve the issue of accurate positioning in a coordinate system. I doubt that regression training would be as easy as classification because how would you know the exact values of your position while recording data.

Unsupervised training would be interesting research stuff in that area. Maybe it's possible to figure out clusters (rooms) automatically.

But simple classification would be a good first step to room occupancy, which might be enough for most use cases.

Another idea that I read somewhere here or in the ESPresense issues: We could try to utilize other stationary devices like Echo Dots to improve the rssi readings of nearby placed ESPs to basically figure out the noise levels.
Could just be an additional dimension in the supervised classification approach.

**Comment by DTTerastar** [comment]
Yes, this wouldn't try and do any coordinate system.  Just the "room" it's in.
